### PR TITLE
AP_GPS: Unix epoch does not have leap seconds

### PR DESCRIPTION
**Pull request by amilcarlucas**
The 15 was probably meant to convert it to UTC, but UTC changes every couple of years. Hardcoding it is just plain wrong.

Unix epoch does not have this problem, no leaps, no deltas, it's just right.

### CODE REVIEW

**Comment by khancyr** [comment]
but gps time take into account the leap seconds no? so you need to remove them to get the right unix time no ?
Otherwise there were a leap second recently some it is maybe time to udapte that if needed

**Comment by amilcarlucas** [comment]
No, GPS time does not have leap seconds
Unix epoch does not have leap seconds either
To convert between these two is straightforward.

The confusion is that UTC does have leap seconds, but UTC is not used in the code !!!

**Comment by OXINARF** [comment]
I'll let the GPS maintainer decide, but:

- Unix epoch is a date in time when leap seconds didn't even exist
- the realtime clock in Unix isn't guaranteed to have leap seconds or not: actually it's definition talks about UTC, which, as you said, have leap seconds. If you take a computer that updates time with a NTP client, the realtime clock is going to go back at some point after a leap second is introduced
- this function says it calculates seconds since Unix epoch, but it doesn't mention if it includes leap seconds or not: in my perspective it should as it looks like it is used to get real (UTC) time

**Comment by amilcarlucas** [comment]
https://en.wikipedia.org/wiki/Unix_time

https://currentmillis.com/

"Unix time" aka "Unix epoch" does not have leap seconds.

**Comment by WickedShell** [comment]
I'm split on this. The reason that I'm sure the -15 is there, is that the ublox driver is requesting UTC time from the GPS unit, rather then GPS time.

Frankly I'm not sure what the better approach is on a systematic level, the ublox driver currently pulls from the iTOW which is based upon the selected navigational epoch, and may or may not be valid yet. We can explicitly request a time message in order to observe if it is infact valid (and the M8P patch which introduced the use of PVT can monitor it for free). However if we detect an invalid UTC conversion I'm really not sure what to do with it, as waiting up to 12.5 minutes for the data is unacceptable.

We can't swap to just GPS based either, as when the user is using a different nav setup (IE Galileo/GLONASS/Beidou) then we have the same time accuracy problems as we do now.

Basically I think this is an acceptable correction for the moment, at the same time though I want to emphasize that this isn't a reliable fix by any means for synchronization on the unix epoch.

**Comment by amilcarlucas** [comment]
Acording to http://www.navipedia.net/index.php/Time_References_in_GNSS only GLONASS has leap seconds:

" Unlike GPS, Galileo or BeiDou, GLONASS time scale implements leap seconds, like UTC "

Waiting 12,5 Minutes for a correct leap second is a long time. :)

**Comment by WickedShell** [comment]
Yes, but again you aren't guaranteed to be using glonass. I'm just pointing
out that it is an entirely unresolved issue.

**Comment by amilcarlucas** [comment]
Hallo Tom,

GPS time has no leap seconds - Unix time has no leap seconds -> hence no need to subtract 15
Galileo time has no leap seconds - Unix time has no leap seconds -> hence no need to subtract 15
Beidou time has no leap seconds - Unix time has no leap seconds -> hence no need to subtract 15

GLONASS time has leap seconds <- this is the only one that can cause a conversion problem.
All other are unambiguous. GPS can optionally provide correct UTC time (with leap seconds),
provided you receive the leap seconds submessage. That can only be guaranteed if you listen to GPS for more than 12,5 Minutes.

Tom Please explain the "you aren't guaranteed to be using glonass" statement. I do not get that point.

**Comment by WickedShell** [comment]
@amilcarlucas I'm Michael :)

I'm aware of which ones have leap seconds. My point is that if you have configured for any other navigational base other then GPS this function is already broken by a lot more then 15-18 seconds. Galileo and Beidou have their own independent epochs and are already not converting correctly.

There is currently no catch all for us as we allow the user to choose to use one of the other systems exclusively if they wish, in which case we can be missing the required conversion information already. All I'm trying to warn you of, is that depending upon how your u-blox is configured this will still be converting incorrectly.

**Comment by amilcarlucas** [comment]
Thanks for the explanation Michael. Your point is now clear to me. Thank you

**Comment by OXINARF** [comment]
@amilcarlucas I'm glad you shared that Wikipedia article since it contradicts itself. It says in the beginning that it doesn't count leap seconds, but then you read the section about leap seconds and it shows that indeed the majority of systems count leap seconds, even if Unix time doesn't have a way to show it - it just repeats the number, making it a non-strictly increasing function. It doesn't handle leap seconds, but it counts them.

As I said, this function tries, in my perspective, to get time in UTC so it should remove leap seconds from a GPS time - which are 18 now though.

@WickedShell Up to you to decide on this.

**Comment by tridge** [comment]
perhaps we should have a GPS_LEAP_SECONDS parameter. It would auto-update when the driver has the information available, and persist between boots. It would default to the current value. The default could even be auto-set by the build system if we want that.

**Comment by amilcarlucas** [comment]
@OXINARF please read the article carefully. It describes *flawed* Unix time implementations where leap seconds are taken into account, when they should have not. Like the ardupilot implementation. :)

According the the Mavlink documentation and many of the comments in the code, it should deliver Unix time.

@tridge your suggestion is a good one, it would solve a lot of the issues. But it requires changing the documentation and touching a lot of code. If ALL documentation and code says UTC then there is no confusion anymore. And that solution should take into account the different GPS epoch of Beidou, Galileo and GLOSSNAS systems.

For applications where multiple different (heterogeneous) embedded systems need to be in sync Unix time is a superior solution, It just works all the time, even when the seconds do leap (June and December)!!!
I will keep the patch in my local code and continue to use Linux time.

**Comment by OXINARF** [comment]
@amilcarlucas They aren't flawed, they follow the spec... For example, 1483228800 refers to what? *2016-12-31T23:59:60(UTC)* **or** *2017-01-01T00:00:00(UTC)*. What you are talking about is like number of seconds since *1970-01-01T00:00:10(TAI)* and that is breaking the spec. Here is a good page about it: http://www.madore.org/~david/computers/unix-leap-seconds.html

Anyway, I've talked with @WickedShell and I think we reached the best option. He'll write about it later.

**Comment by WickedShell** [comment]
The conclusion I reached from talking with @OXINARF is that the best approach is to update the appropriate number of leap seconds to remove as a hardcoded value. Then convert the u-blox driver to request GPS time instead of UTC time (which is what we are currently asking for).

There are still a couple of sources of error that are possible with this approach:
1. The number of leap seconds has changed since the value was hard coded
2. We have requested a time base from the GPS that doesn't include enough conversion information because we don't yet know how to convert the time base.

I don't think either of these are super important however as your maximum error is still 18 seconds which isn't huge for this, most UNIX systems have clocks that are easily that far off already. The general case is much better, and we need to remember that other GPS users need this to be well behaved which means that in general we should be subtracting the leap seconds.

I don't think a parameter system is worth the complexity though, as it adds a lot of driver over head for a very small win. 

A couple of cases that aren't covered is that conversions in this function are already very broken if you are operating with just Galileo or Beidou for example and using that as your time base, and since we may not be allowed to reconfigure the GPS we have to accept that this is a possible situation, so even with a parameter this would still be a problem.

**Comment by rmackay9** [comment]
I'm not completely sure but I think the only place we wanted to use UTC time was in the nav-delay mission command which allows a user to set a specific time that a command will begin (or more accurately it delays the mission until that time arrives).  When testing this command, I've noticed the time is off from UTC by about 20seconds or so.  To get around that I made it print how many seconds until the nav-delay completes so the user can keep an eye on it.
It's a rarely used command so I don't think we should worry about it's accuracy too much.
